doc/sphinx/language/core/basic.rst

```diff
@@ -0,0 +1,509 @@
+  represent a double quote character within a string.  In the grammar,
+  strings are identified with :production:`string`.
+
+Keywords
+  The following character sequences are reserved keywords that cannot
+  be used as identifiers::
+
+    _ Axiom CoFixpoint Definition Fixpoint Hypothesis IF Parameter Prop
+    SProp Set Theorem Type Variable as at by cofix discriminated else
+    end exists exists2 fix for forall fun if in lazymatch let match
+    multimatch return then using where with
+
+  Note that notations and plugins may define additional keywords.
+
+Other tokens
+  The set of tokens defined at any given time can vary because the
+  :cmd:`Notation` command can define new tokens.  A :cmd:`Require`
+  command may load more notation definitions, while the end of a
+  :cmd:`Section` may remove notations.  Some notations are defined in
+  the standard library (see :ref:`thecoqlibrary`) and are generally
+  loaded automatically at startup time.
+
+  Here are the character sequences that |Coq| directly defines as tokens
+  without using :cmd:`Notation`::
+
+    ! #[ % & ' ( () (bfs) (dfs) ) * ** + , - ->
+    . .( .. ... / : ::= := :> :>> ; < <+ <- <:
+    <<: <= = => > >-> >= ? @ @{ [ [= ] _
+    `( `{ { {| | |- || }
+
+  When multiple tokens match the beginning of a sequence of
+  characters, the longest matching token is used.  Occasionally you
+  may need to insert spaces to separate tokens.  For example, if ``~``
+  and ``~~`` are both defined as tokens, the inputs ``~ ~`` and ``~~``
+  generate different tokens, whereas if `~~` is not defined, then the
+  two inputs are equivalent.
+
+Essential vocabulary
+--------------------
+
+This section presents the most essential notions to understand the
+rest of the |Coq| manual: :term:`terms <term>` and :term:`types
+<type>` on the one hand, :term:`commands <command>` and :term:`tactics
+<tactic>` on the other hand.
+
+.. glossary::
+
+   term
+
+     The base expressions of the |Coq| language are terms.  Terms can
+     represent computable programs, program types, mathematical
+     propositions and mathematical proofs.
+
+     The syntax of terms is as follows.  Each one of the listed
+     constructs is presented in a dedicated section.  Some of these
+     constructs (like :n:`@term_product_or_fun`) are part of the core
+     language that the kernel of Coq understands and are therefore
+     described in :ref:`the present chapter <core-language>`, while
+     some others (like :n:`@term_if`) are language extensions that are
+     presented in :ref:`the next chapter <extensions>`.
+
+     .. insertprodn term qualid_annotated
+
+     .. prodn::
+        term ::= @term_product_or_fun
+        | @term_let
+        | @term_if
+        | @term_fix
+        | @term_cofix
+        | @term100
+        term100 ::= @term_cast
+        | @term10
+        term10 ::= @term_application
+        | @one_term
+        one_term ::= @ @qualid_annotated
```

Comment:
An index of nonterminals is easy to produce.  doc_grammar could also generate a list of symbols to go in an index (since many symbols are used in multiple productions, they'd have to be qualified by the name of the nonterminal using the symbol).  I could do that after we're done updating syntax in the doc.

OTOH, repeating the grammar in the appendix is almost as good, less work and helpful in some other ways.  Users can use the browser search to find things.

Reply:
You might be right (but then the repeated productions would have to link back to the place where they are documented). OTOH having an index (with the qualified disambiguation like you said) would show instantly to the user that there are several rules corresponding to same terminal (typical example: `match`). Whereas some users might search for `match` in the full grammar and not realize that the first instance is not the only occurrence.

Reply:
My browser, Chrome, shows the number of matches whenever you do a text search, so it's easy to see that there are multiple occurrences.  I don't know if most browsers do that.

Reply:
I think so. But it is easy to miss if you don't realize that there might be several matches.